### DISCUSSION TITLE
Multiple disallowed versions are not merged

### DISCUSSION THREAD

**Discussion by bergerst** [renovatebot/renovate#15344] [Request Help] [answered]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

After defining my own disallowed versions for packages, Renovate creates PRs for commons-collections version 20040616 even though there's a global workaround `mavenCommonsAncientVersion` disallowing any version starting with `200\\d{5}`.

When I remove my own packageRules, the PR for updating commons-collections is closed. When I copy paste the workaround into the renovate.json, the PR is also closed.

It seems like multiple disallowed versions are not merged but instead overwrite each other. I think #14817 is related to this bug.

See https://github.com/bergerst/renovatebot_test/ for a reproduction repository.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Looking up commons-io:commons-io in repository https://repo.maven.apache.org/maven2/
DEBUG: Looking up log4j:log4j in repository https://repo.maven.apache.org/maven2/
DEBUG: Looking up commons-collections:commons-collections in repository https://repo.maven.apache.org/maven2/
DEBUG: Found 21 new releases for commons-io:commons-io in repository https://repo.maven.apache.org/maven2/
DEBUG: Found 14 new releases for log4j:log4j in repository https://repo.maven.apache.org/maven2/
DEBUG: Found 17 new releases for commons-collections:commons-collections in repository https://repo.maven.apache.org/maven2/
DEBUG: Package releases lookups complete
{
  "baseBranch": "main"
}
DEBUG: branchifyUpgrades
DEBUG: 2 flattened updates found: commons-io:commons-io, commons-collections:commons-collections
DEBUG: Returning 2 branch(es)
DEBUG: Fetching changelog: https://gitbox.apache.org/repos/asf?p=commons-io (2.8.0 -> 2.11.0)
INFO: Unknown platform, skipping changelog fetching.
{
  "sourceUrl": "https://gitbox.apache.org/repos/asf?p=commons-io",
  "hostType": null
}
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "config": {
    "maven": [
      {
        "datasource": "maven",
        "packageFile": "pom.xml",
        "deps": [
          {
            "datasource": "maven",
            "depName": "commons-io:commons-io",
            "currentValue": "2.8.0",
            "fileReplacePosition": 565,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2"
            ],
            "depType": "compile",
            "depIndex": 0,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "2.11.0",
                "newValue": "2.11.0",
                "releaseTimestamp": "2021-07-10T01:43:00.000Z",
                "newMajor": 2,
                "newMinor": 11,
                "updateType": "minor",
                "branchName": "renovate/commons-io-commons-io-2.x"
              }
            ],
            "warnings": [],
            "versioning": "maven",
            "sourceUrl": "https://gitbox.apache.org/repos/asf?p=commons-io",
            "homepage": "https://commons.apache.org/proper/commons-io/",
            "currentVersion": "2.8.0",
            "isSingleVersion": true,
            "fixedVersion": "2.8.0"
          },
          {
            "datasource": "maven",
            "depName": "log4j:log4j",
            "currentValue": "1.2.17",
            "fileReplacePosition": 725,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2"
            ],
            "depType": "compile",
            "depIndex": 1,
            "updates": [],
            "warnings": [],
            "versioning": "maven",
            "sourceUrl": "http://svn.apache.org/viewvc/logging/log4j/tags/v1_2_17_rc3",
            "homepage": "http://logging.apache.org/log4j/1.2/",
            "currentVersion": "1.2.17",
            "fixedVersion": "1.2.17"
          },
          {
            "datasource": "maven",
            "depName": "commons-collections:commons-collections",
            "currentValue": "3.2.2",
            "fileReplacePosition": 914,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2"
            ],
            "depType": "test",
            "depIndex": 2,
            "updates": [
              {
                "bucket": "major",
                "newVersion": "20040616",
                "newValue": "20040616",
                "releaseTimestamp": "2005-11-08T22:31:31.000Z",
                "newMajor": 20040616,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/commons-collections-commons-collections-20040616.x"
              }
            ],
            "warnings": [],
            "versioning": "maven",
            "currentVersion": "3.2.2",
            "isSingleVersion": true,
            "fixedVersion": "3.2.2"
          }
        ],
        "packageFileVersion": "1.0-SNAPSHOT"
      }
    ]
  }
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment, marked as answer]
This is intended behaviour as the field isn't mergeable

**Reply by bergerst** [reply]
I'm not really satisfied with this answer. Right now global disallowed versions like in my repo lead to all maven workarounds to be disabled.

The current behavior is neither intuitive, documented or practical.

Right now I would have to restrict my atlassian/redhat/... filter to package names in order to enable the workarounds again.

**Reply by rarkins** [reply]
Unfortunately, that doesn't really matter and definitely doesn't change what the answer is

**Reply by HonkingGoose** [reply]
> The current behavior is neither intuitive, documented or practical.

Let's see if I can at least document the behavior then. 😉 

> This is intended behaviour as the field isn't mergeable

@rarkins Do you mean that `allowedVersions` is not mergeable? The [docs for `allowedVersions`](https://docs.renovatebot.com/configuration-options/#allowedversions) doesn't mention `mergeable`:

> ### allowedVersions
> 
> A version range or regex pattern capturing allowed versions for dependencies.
> 
> <html><body>
> 
> 
> Name | Value
> -- | --
> type | string
> parent | packageRules
> default | `null`
> 
> 
> </body>
> </html>

Should I edit the code below so `mergeable` is set to `false`?

https://github.com/renovatebot/renovate/blob/d4541ec4293cc1d429f440a1fd1dcbb298d468a8/lib/config/options/index.ts#L1128-L1137

**Reply by rarkins** [reply]
String fields are never mergeable

**Reply by HonkingGoose** [reply]
Oh okay! In that case, `mergeable` should be `false` for all `string` fields. Putting that in the code is going to be error prone.

How about we put something like this in our docs generating code?

```
IF type === 'string'
THEN insert new row `mergeable` = `false` in the table for the config option
```

**Reply by rarkins** [reply]
I think in most cases that's just going to only add noise to our docs with little incremental value. "of course" strings aren't mergeable - it's not like people should expect us to concatenate them and it would somehow work. e.g. adding mergeable=false to the `branchPrefix` entry is not adding any value

**Reply by HonkingGoose** [reply]
> I think in most cases that's just going to only add noise to our docs with little incremental value.

The docs should be high signal, low noise. 😄 

> "of course" strings aren't mergeable - it's not like people should expect us to concatenate them and it would somehow work. e.g. adding mergeable=false to the `branchPrefix` entry is not adding any value

To me this was not immediately obvious. It might be good to add an admonition to our config option docs, see PR:

- #15357

**Reply by viceice** [reply]
Maybe better document that only settings with `mergeable=true` are mergable, all others are not. I don't like to add it to all tables because of the already mentioned noise and blowup

- #15243

**Reply by HonkingGoose** [reply]
I created a PR to add this note (#15357):

```markdown
<!-- prettier-ignore -->
!!! note
    Config options with `type=string` are always non-mergeable, so `mergeable=false`.
```

Do you want a message like this then?

```markdown
<!-- prettier-ignore -->
!!! note
    Only config options where `mergeable=true` can be merged.
    If the `mergable` field is absent in the docs then assume the config option is non-mergeable.
```
